snuba/datasets/join_schema.py

```diff
@@ -0,0 +1,92 @@
+from dataclasses import dataclass
+from enum import Enum
+from typing import Callable, Mapping, Optional, Sequence
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.datasets.schema import Schema
+
+
+class JoinType(Enum):
+    INNER = "INNER"
+    LEFT = "LEFT"
+    RIGHT = "RIGHT"
+    FULL = "FULL"
+
+
+@dataclass(frozen=True)
+class JoinMapping():
+    left_alias: str
+    left_column: str
+    right_alias: str
+    right_column: str
+
+    def get_where_clause(self) -> str:
+        return f"{self.left_alias}.{self.left_column} = " \
+            f"{self.right_alias}.{self.right_column}"
+
+
+@dataclass(frozen=True)
+class JoinedSource():
+    source: Schema
+    alias: Optional[str]
+
+
+@dataclass(frozen=True)
+class JoinStructure:
+    """
+    Abstracts the join clause as a tree.
+    Every node in the tree is either a join itself or a
+    schema with an alias.
+    Traversing the tree it is possible to build the join
+    clause.
+
+    This does not validate the join makes sense nor it checks
+    the aliases are valid.
+    """
+    left_schema: JoinedSource
+    right_schema: JoinedSource
+    mapping: Sequence[JoinMapping]
+    join_type: JoinType
+
+    def get_where_clause(self) -> str:
```

Comment:
This appears to be a from clause, instead of the where clause. Is that right?

Reply:
yes of course ..... I spent hours to find the right naming and then I wrote the wrong one.